### PR TITLE
Add "garden publish" command

### PR DESCRIPTION
**Pull request by WillEngler**
Implements the bulk of #53.

## Overview

This PR lays down most of the work to implement the `garden-ai garden publish` command. It also adds `garden-ai garden add-pipeline`. Publishing a garden creates a new record in Globus Search for that garden, including any child pipelines.

There are some refactoring changes too:
- Local "db" operations are now in a separate module instead of being methods on GardenClient.
- Some old first drafts of publication have been deleted or repurposed.

## Discussion

There are a few smells and TODOs here. The biggest smell is using JSON metadata dictionaries instead of Garden and Pipeline models. As I note in [this comment](https://github.com/Garden-AI/garden/issues/96#issuecomment-1505451367), we can't currently go from serialized garden or pipeline JSON back to Garden or Pipeline Pydantic models. We'll need to address that soon, likely in this sprint. It seemed like overkill to work that out in this PR, but we should go back and clean this up soon.

There are some TODOs remaining:
- ~There was a backend error on DOI minting for Gardens so I punted on that for this PR. I intend to address it in a follow-up PR this sprint.~ [Figured it out. Included it.](https://github.com/Garden-AI/garden/pull/109/commits/c0a873b80a4bf973727363b2cea25b9d1b283e2b)
- We'd like to be able to refer to gardens and pipelines by DOI. Maybe even title. I intend to include selecting by DOI in a follow-up PR.
- ~I need to make the dev Search index publicly writable or at least add all the Garden devs to it. (I couldn't write to the one Eric set up but I had an endpoint lying around that I subbed in.)~ @ericblau  added me as admin to the old index and I invited the Garden devs to it. We can't make it publicly writeable (which is a good and sane constraint) so we'll just have to invite people to the "Garden Dev Search" Globus group until we move ingest to the backend.

## Testing

Two new CLI tests exercise the new commands. Running end-to-end by invoking the CLI, I was able to confirm that the JSON I intended to upload to the search index got there with `globus-search query 8386583d-e1cf-4b2f-ae66-b335ab5c9a21 "*"`

----
:books: Documentation preview :books:: https://garden-ai--109.org.readthedocs.build/en/109/

### CODE REVIEW

File: garden_ai/local_data.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
+import json
```
Comment (WillEngler) [inline garden_ai/local_data.py:1]: This is almost all pulled directly from client.py, just stripping out `self`.

File: garden_ai/app/garden.py [hunk 2]
```diff
@@ -168,9 +174,119 @@ def create(
...
+
+
+# Right now we can make a Garden model from Garde JSON iff it has no pipelines.
+# Make a pipeline-less copy of the JSON so that we can make a Garden model
+# and use its DOI generating method.
+# TODO: clean up this hack once we fix Pydantic model deserialization.
+def mint_doi_from_garden_meta(garden_meta: Dict, client: GardenClient):
+    garden_copy = deepcopy(garden_meta)
+    garden_copy["pipelines"] = []
+    # No DOI currently serializes to None, which is not accepted by the Garden model.
```
Comment (OwenPriceSkelly) [inline garden_ai/app/garden.py:256]: if the 'doi' key isn't in the `garden_copy` dict, there should be no errors (but this is p surprising behavior, we can/should revisit how "required" a doi should be for v0)

**Review by WillEngler — COMMENTED** [review]
Notes for reviewers

File: garden_ai/app/garden.py [hunk 3]
```diff
@@ -168,9 +174,119 @@ def create(
...
+    if not garden_copy["doi"]:
+        garden_copy["doi"] = ""
```
Comment (OwenPriceSkelly) [inline garden_ai/app/garden.py:258]:
```suggestion
        del garden_copy["doi"]
```

File: garden_ai/app/garden.py [hunk 4]
```diff
@@ -168,9 +174,119 @@ def create(
...
+    garden_model = Garden(**garden_copy)
+    doi = client._mint_doi(garden_model)
+    return doi
+
+
+def get_pipeline_meta(pipeline_uuid: str):
+    maybe_pipeline = local_data.get_local_pipeline(pipeline_uuid)
+    if not maybe_pipeline:
+        logger.fatal(f"Could not find garden with uuid {pipeline_uuid}")
```
Comment (OwenPriceSkelly) [inline garden_ai/app/garden.py:267]:
```suggestion
        logger.fatal(f"Could not find pipeline with uuid {pipeline_uuid}")
```

**Review by OwenPriceSkelly — APPROVED** [review]
looks good to me! one typo but otherwise zero complaints 🙌

**Comment by WillEngler** [comment]
Thanks! Just committed the suggestions you made.

File: garden_ai/local_data.py
[the remark this answers is not in the document]
Comment (WillEngler) [reply garden_ai/local_data.py:1]: Remote storage should just supplant local storage imo. The client will have the auth info for how to talk to our backend, so calls out to the API endpoints that hit the remote DB will involve the client in some form.

> are we going to need to pull some of this functionality back into the client?

Basically no. All this logic should be in the backend routes that wrap database calls.

**Pull request merged**